### PR TITLE
Add option to run command on bell

### LINKED ISSUE #1528: Add support for audio bell
Which operating system does the issue occur on?
macOS High Sierra v10.13.6

I don't get any audio notification for the terminal bell. The config file has options for `visual_bell`, but that's of no help when I'm looking at other fullscreen windows.

For the time being, I've worked around the problem using tmux:
```tmux
# Some hacks for OSX Alacritty audio bell
set -g monitor-bell on
set -g bell-action other
set-hook -g alert-bell 'run-shell "afplay /System/Library/Sounds/Funk.aiff"'
```

Does Alacritty not provide an audio bell? If not, is the feature planned or out of scope?

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes #1528.

### CODE REVIEW

**Comment by kchibisov** [comment]
The visual bell in `term/mod.rs` was kept that way intentionally, since it's actually a visual bell.

**Comment by kchibisov** [comment]
Configuration follows https://github.com/alacritty/alacritty/pull/2298#pullrequestreview-225782173

File: alacritty.yml [hunk 1]
```diff
@@ -245,31 +245,46 @@
...
+  # This program is executed whenever the bell is rung.
+  #
+  # When set to `command: None`, bell command will be disabled completely.
+  #
+  # Example:
+  #   command:
+  #     program: notify-send
+  #     args: ["Hello, World!"]
+  #
+  #command: None
```
Comment (chrisduerr) [inline alacritty.yml:287]:
```suggestion
  #program: None
```

File: alacritty_terminal/src/config/mod.rs [hunk 2]
```diff
@@ -115,6 +115,10 @@ pub struct Config<T> {
     #[serde(skip)]
     pub hold: bool,
 
+    // TODO: DEPRECATED
+    #[serde(default, deserialize_with = "failure_default")]
+    pub visual_bell: Option<BellConfig>,
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/mod.rs:120]:
```suggestion
    visual_bell: Option<BellConfig>,
```

File: alacritty_terminal/src/config/visual_bell.rs [hunk 3]
```diff
@@ -2,47 +2,73 @@ use std::time::Duration;
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/visual_bell.rs:2]: This file needs to be deleted I think?

File: alacritty_terminal/src/term/mod.rs [hunk 4]
```diff
@@ -553,10 +553,10 @@ fn cubic_bezier(p0: f64, p1: f64, p2: f64, p3: f64, x: f64) -> f64 {
 
 impl VisualBell {
     pub fn new<C>(config: &Config<C>) -> VisualBell {
-        let visual_bell_config = &config.visual_bell;
+        let bell_config = config.bell();
```
Comment (chrisduerr) [inline alacritty_terminal/src/term/mod.rs:556]: Could we maybe implement `From<Bell> for VisualBell`?

File: alacritty_terminal/src/config/bell.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    #[serde(deserialize_with = "failure_default")]
+    pub animation: BellAnimation,
+
+    /// Visual bell duration in milliseconds.
+    #[serde(deserialize_with = "failure_default")]
+    duration: u16,
+
+    // FIXME RGB deserialization always successedes and uses #000000 color. However it should
+    // fallback to a proper color. So it should be reworked either with
+    // https://github.com/alacritty/alacritty/pull/3507 or after that PR will land.
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:25]: I mean that you'll start doing manual deserialization, right? So it'll be possible to do so as well here, since Rgb will be free of hardcode to `#000000`. 

**NOTE** I don't introduce new bug here, since it was like that already.

File: alacritty_terminal/src/config/bell.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    /// Visual bell flash color.
+    #[serde(deserialize_with = "failure_default")]
+    pub color: Rgb,
+
+    /// Command to run on bell.
+    #[serde(deserialize_with = "deserialize_command")]
+    pub command: Option<Program>,
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:32]: As discussed on IRC we'll keep `command` since it's used in bindings already, so it fits default config more.

File: alacritty_terminal/src/config/mod.rs (hunk 2 above)
Comment (kchibisov) [reply alacritty_terminal/src/config/mod.rs:120]: It's required to throw deprecation warning.

File: alacritty_terminal/src/config/bell.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    // FIXME RGB deserialization always successedes and uses #000000 color. However it should
+    // fallback to a proper color. So it should be reworked either with
+    // https://github.com/alacritty/alacritty/pull/3507 or after that PR will land.
+    /// Visual bell flash color.
+    #[serde(deserialize_with = "failure_default")]
+    pub color: Rgb,
+
+    /// Command to run on bell.
+    #[serde(deserialize_with = "deserialize_command")]
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:31]: `Program` doesn't have `Default` and I can't impl it, before there's none logically, so can't do that.

File: alacritty_terminal/src/config/bell.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisduerr) [reply alacritty_terminal/src/config/bell.rs:25]: Not for all fields, that seems to be pretty much impossible since we have 16 colors. So having 16 different deserialization functions just doesn't make any sense.

[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:25]: Hm, so how should I handle deserialization of Rgb then?

File: CHANGELOG.md [hunk 8]
```diff
@@ -24,6 +24,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Font fallback on Windows
 - Support for Fontconfig embolden and matrix options
 - Opt-out compilation flag `winpty` to disable WinPTY support
+- Option to run command on bell which can be set in `bell.command`
```
Comment (chrisduerr) [inline CHANGELOG.md:27]: We should also add a change log entry for moving visual_bell option to bell.

File: alacritty/src/event.rs [hunk 9]
```diff
@@ -758,6 +759,15 @@ impl<N: Notify + OnResize> Processor<N> {
         processor.ctx.terminal.dirty = true;
     }
 
+    fn start_program(program: &Program) {
+        let args = program.args().to_vec();
+
+        match start_daemon(program.program(), &args) {
+            Ok(_) => debug!("Launched {} with args {:?}", program.program(), args),
+            Err(_) => warn!("Unable to launch {} with args {:?}", program.program(), args),
+        }
```
Comment (chrisduerr) [inline alacritty/src/event.rs:768]: We have this construct in a lot of places, should it maybe be moved to `start_daemon`?

If it's used just for visual bell, it probably shouldn't be `start_program`, but I see nothing particularly visual bell specific here.

File: alacritty_terminal/src/config/bell.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    pub animation: BellAnimation,
+
+    /// Visual bell duration in milliseconds.
+    #[serde(deserialize_with = "failure_default")]
+    duration: u16,
+
+    // FIXME RGB deserialization always successedes and uses #000000 color. However it should
+    // fallback to a proper color. So it should be reworked either with
+    // https://github.com/alacritty/alacritty/pull/3507 or after that PR will land, since it'll
+    // allow to impl manual deserialization properly.
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:26]: Adding a reminder that this needs to be done before the PR can be merged.

File: alacritty_terminal/src/config/bell.rs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+{
+    // Deserialize to generic value.
+    let val = Value::deserialize(deserializer)?;
+
+    // Accept `None` to disable the bell command.
+    if val.as_str().filter(|v| v.to_lowercase() == "none").is_some() {
+        return Ok(None);
+    }
+
+    match <Option<Program>>::deserialize(val) {
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:67]:
```suggestion
    match Program::deserialize(val) {
```

File: alacritty_terminal/src/config/bell.rs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+impl BellConfig {
+    /// Visual bell duration in milliseconds.
+    #[inline]
+    pub fn duration(&self) -> Duration {
+        Duration::from_millis(u64::from(self.duration))
+    }
+}
+
+fn deserialize_command<'a, D>(deserializer: D) -> std::result::Result<Option<Program>, D::Error>
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:55]:
```suggestion
fn deserialize_bell_command<'a, D>(deserializer: D) -> Result<Option<Program>, D::Error>
```

File: alacritty_terminal/src/config/bell.rs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+    // Accept `None` to disable the bell command.
+    if val.as_str().filter(|v| v.to_lowercase() == "none").is_some() {
+        return Ok(None);
+    }
+
+    match <Option<Program>>::deserialize(val) {
+        Ok(command) => Ok(command),
+        Err(err) => {
+            error!(target: LOG_TARGET_CONFIG, "Problem with config: {}; not using anything", err);
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:70]:
```suggestion
            error!(target: LOG_TARGET_CONFIG, "Problem with config: {}; ignoring field", err);
```

File: alacritty/src/event.rs (hunk 9 above)
Comment (kchibisov) [reply alacritty/src/event.rs:768]: We can't abstract, since messages are different everywhere, so you may move, but you should supply the messages, so I feel like it's fine the way it is now.

File: alacritty_terminal/src/config/bell.rs (hunk 10 above)
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:26]: I mean it's not a bug from that PR so I'm not sure how it's related and prevents it from being merged, I've just highlighted a very old issue.

File: alacritty_terminal/src/config/bell.rs (hunk 11 above)
Comment (kchibisov) [reply alacritty_terminal/src/config/bell.rs:67]: The field is an option itself, so either do that or manually using `Ok(Some(command))`, so there's really not much difference, and I'd stick with the current approach, since we're using it in launcher deser already.

**Comment by kchibisov** [comment]
Blocked by #3507

**Comment by kchibisov** [comment]
I've rebased pr and solved some the remaining issue on that PR. The first commit is just rebase, and the last one is resolution of that one issue we had.

File: alacritty/src/event.rs [hunk 14]
```diff
@@ -839,8 +839,15 @@ impl<N: Notify + OnResize> Processor<N> {
...
-                    TerminalEvent::Urgent => {
-                        processor.ctx.window.set_urgent(!processor.ctx.terminal.is_focused)
+                    TerminalEvent::Bell => {
+                        let _ = processor
+                            .ctx
+                            .config
+                            .bell()
+                            .command
+                            .as_ref()
+                            .map(Self::start_bell_program);
```
Comment (chrisduerr) [inline alacritty/src/event.rs:849]: This formatting seems very unfortunate. Could we maybe extract a `let bell_config = processor.ctx.config.bell();` to make this use only two lines? I don't care how it is achieved, but this seems very wasteful. `if let Some(xxx)` would be another option I suppose, since mapping just to execute is kinda meh anyways.

File: alacritty/src/event.rs (hunk 9 above)
Comment (chrisduerr) [reply alacritty/src/event.rs:1067]: Do the messages need to be different? Or would we be fine with just having a uniform message for launching processes and having that  fail?

I feel like it's probably not super necessary to be able to have a custom message for every different way we want to spawn a program?

File: alacritty_terminal/src/config/bell.rs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    pub command: Option<Program>,
+}
+
+impl Default for BellConfig {
+    fn default() -> Self {
+        Self {
+            animation: Default::default(),
+            duration: Default::default(),
+            color: DEFAULT_BELL_COLOR,
+            command: Default::default(),
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:38]:
```suggestion
            command: Default::default(),
            color: DEFAULT_BELL_COLOR,
```

Doesn't matter at all, but this would look so much better. :)

File: alacritty_terminal/src/config/bell.rs (hunk 11 above)
Comment (chrisduerr) [reply alacritty_terminal/src/config/bell.rs:83]: I feel like that probably adds unnecessary overhead though and it doesn't clearly communicate that this cannot fail. While just using `Ok(Some(command))` makes it clear that it can never be `None` at this point.

I feel like `Program::deserialize` is the better choice here.

File: alacritty/src/event.rs [hunk 16]
```diff
@@ -1078,6 +1079,17 @@ impl<N: Notify + OnResize> Processor<N> {
     }
 }
 
+fn start_program<I, S>(program: &str, args: I)
+where
+    I: IntoIterator<Item = S> + Debug + Copy,
+    S: AsRef<OsStr>,
+{
+    match start_daemon(program, args) {
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1087]: Do we not want to provide this functionality globally in `start_daemon`? Are there instances where we do not want to print a warning?

File: alacritty/src/event.rs [hunk 17]
```diff
@@ -839,8 +835,13 @@ impl<N: Notify + OnResize> Processor<N> {
...
                     TerminalEvent::Title(title) => processor.ctx.window.set_title(&title),
                     TerminalEvent::Wakeup => processor.ctx.terminal.dirty = true,
-                    TerminalEvent::Urgent => {
-                        processor.ctx.window.set_urgent(!processor.ctx.terminal.is_focused)
+                    TerminalEvent::Bell => {
+                        let bell_config = processor.ctx.config.bell();
+                        let _ = bell_config
+                            .command
+                            .as_ref()
+                            .map(|program| start_program(program.program(), program.args()));
```
Comment (chrisduerr) [inline alacritty/src/event.rs:843]:
```suggestion
                        let bell_command = processor.ctx.config.bell().command.as_ref();
                        let _ = bell_command.map(|cmd| start_program(cmd.program(), cmd.args()));
```

File: alacritty_terminal/src/config/bell.rs [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+            );
+
+            Ok(DEFAULT_BELL_COLOR)
+        },
+    }
+}
+
+fn deserialize_bell_command<'a, D>(
+    deserializer: D,
+) -> std::result::Result<Option<Program>, D::Error>
```
Comment (chrisduerr) [inline alacritty_terminal/src/config/bell.rs:71]:
```suggestion
) -> Result<Option<Program>, D::Error>
```

File: alacritty/src/event.rs (hunk 16 above)
Comment (kchibisov) [reply alacritty/src/event.rs:1087]: I'm not sure that we should do that in `start_daemon`. It does the one thing, and have a special impl on each platform, so I'm not sure if there's any value in moving it there. Hiding errors from library users isn't something I really like tbqh.

Comment (chrisduerr) [reply alacritty/src/event.rs:1087]: Could you move it out of `alacritty_terminal` then maybe? There's no reason for it to be there.

File: alacritty/src/daemon.rs [hunk 19]
```diff
@@ -1,77 +1,64 @@
...
-where
-    I: IntoIterator<Item = S>,
-    S: AsRef<OsStr>,
-{
     // Setting all the I/O handles to null and setting the
     // CREATE_NEW_PROCESS_GROUP and CREATE_NO_WINDOW has the effect
     // that console applications will run without opening a new
     // console window.
-    Command::new(program)
+    #[cfg(windows)]
```
Comment (chrisduerr) [inline alacritty/src/daemon.rs:50]: Feels like it would be easier to have these as separate methods, even if they're not exposed publicly? Like this the annotations are quite complicated to keep track of since they're relatively big.

Comment (kchibisov) [reply alacritty/src/daemon.rs:50]: ok.

File: alacritty/src/daemon.rs [hunk 20]
```diff
@@ -1,34 +1,53 @@
...
+/// Start the daemon and log error on failure.
+pub fn start_daemon<I, S>(program: &str, args: I)
+where
+    I: IntoIterator<Item = S> + Debug + Copy,
+    S: AsRef<OsStr>,
+{
+    match start_daemon_impl(program, args) {
+        Ok(_) => debug!("Launched {} with args {:?}", program, args),
+        Err(_) => warn!("Unable to launch {} with args {:?}", program, args),
+    };
```
Comment (chrisduerr) [inline alacritty/src/daemon.rs:24]:
```suggestion
    }
```

File: alacritty/src/daemon.rs [hunk 21]
```diff
@@ -1,34 +1,53 @@
...
+        .spawn()
+        .map(|_| ())
 }
 
 #[cfg(not(windows))]
-pub fn start_daemon<I, S>(program: &str, args: I) -> io::Result<()>
+fn start_daemon_impl<I, S>(program: &str, args: I) -> io::Result<()>
 where
-    I: IntoIterator<Item = S>,
+    I: IntoIterator<Item = S> + Copy,
```
Comment (chrisduerr) [inline alacritty/src/daemon.rs:50]: Why did you add the `Copy` constraint?

File: alacritty/src/daemon.rs [hunk 22]
```diff
@@ -1,34 +1,53 @@
...
+{
+    match start_daemon_impl(program, args) {
+        Ok(_) => debug!("Launched {} with args {:?}", program, args),
+        Err(_) => warn!("Unable to launch {} with args {:?}", program, args),
+    };
+}
 
-    pub use std::thread::*;
+#[cfg(windows)]
+fn start_daemon_impl<I, S>(program: &str, args: I) -> io::Result<()>
```
Comment (chrisduerr) [inline alacritty/src/daemon.rs:28]:
```suggestion
fn spawn_daemon<I, S>(program: &str, args: I) -> io::Result<()>
```

I think adding `_impl` is a bit strange. A similar but not identical name should do.

File: alacritty_terminal/src/config/bell.rs (hunk 18 above)
Comment (chrisduerr) [reply alacritty_terminal/src/config/bell.rs:71]: +1

File: alacritty/src/daemon.rs (hunk 21 above)
Comment (kchibisov) [reply alacritty/src/daemon.rs:50]: Because you should copy a reference? Otherwise it's not working, since you should print it as well and to pass it somewhere else?
